google/cloud/bigtable/batcher.py

```diff
@@ -139,8 +228,30 @@ def flush(self):
             :dedent: 4
 
         """
-        if len(self.rows) != 0:
-            self.table.mutate_rows(self.rows)
-            self.total_mutation_count = 0
-            self.total_size = 0
-            self.rows = []
+        rows_to_flush = []
+        while not self._rows.empty():
+            rows_to_flush.append(self._rows.get())
+        self.flush_rows(rows_to_flush)
+
+    def flush_async(self):
+        """Sends the current batch to Cloud Bigtable asynchronously."""
+
+        rows_to_flush = []
+        while not self._rows.empty():
+            rows_to_flush.append(self._rows.get())
+        self._executor.submit(self.flush_rows, rows_to_flush)
+
+    def flush_rows(self, rows_to_flush=None):
+        """Mutate the specified rows."""
+        if len(rows_to_flush) > 0:
+            self.table.mutate_rows(rows_to_flush)
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """Clean up resources. Flush and shutdown the ThreadPoolExecutor."""
+        self.close()
+
+    def close(self):
+        """Clean up resources. Flush and shutdown the ThreadPoolExecutor."""
+        self._is_open = False
+        self.flush()
```

Comment:
how are errors handled? the java client throw errors on the failed elements onClose(). Will the batcher fail silently if some elements failed to apply?

Reply:
It uses the existing retry mechanism in `table.mutate_rows`.

https://github.com/googleapis/python-bigtable/blob/main/google/cloud/bigtable/table.py#L680

Reply:
So `table.mutate_rows` returns the response statuses to indicate errors.
We can capture the status codes and raise error.

Reply:
An error is now raised if any of the mutation errored out.